src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.ProjectState.cs

```diff
@@ -199,7 +199,8 @@ public async ValueTask SaveToInMemoryStorageAsync(Project project, DiagnosticAna
                 var serializerVersion = result.Version;
                 foreach (var documentId in result.DocumentIds)
                 {
-                    var document = project.GetTextDocument(documentId);
+                    var document = project.GetTextDocument(documentId)
+                        ?? project.TryGetSourceGeneratedDocumentForAlreadyGeneratedId(documentId);
```

Comment:
Any reason we can't use the async form? I'm guessing the idea here is we only got this far if had ran a generation step earlier but the null checking below makes me wonder if there's other problems this code needs to be durable against.

Reply:
➡️ This one is now documented

Reply:
I guess the statement doesn't seem to be strong enough -- it's not just a matter that it ran somewhere, we have to have ran generators for that _that exact snapshot_. But note that by the time you can make that statement, just calling the async API is fine since you're already returning cached results at that point.

Reply:
⏱️ Will improve